docs/usability/creating_invocable.md

```diff
@@ -0,0 +1,685 @@
+- _GetActiveScopes_ is needed to init the SDE. This cause different plugins to load or not.
+- _GetWorkspaceRoot_ gets the folder that is the root of your workspace.
+- _GetPackagePaths_ gets the folder locations that you will resolve EDK2 paths against.
+- _GetName_ is the thing we are building
+- _AddCommandLineOptions_ gives our settings the chance to set items in the parser object
+- _RetrieveCommandLineOptions_ gives us the chance to read the arguments from the commandline
+
+Now that we have our base methods, let's add one more to control the configurations we are going to build.
+
+```python
+
+class BinaryBuildSettingsManager():
+    ''' Platform settings will be accessed through this implementation. '''
+
+    ....
+
+    def GetConfigurations(self):
+        '''
+        Gets the next configuration of this run
+        This is a generator pattern - use yield
+        '''
+        raise NotImplementedError()
+```
+
+- _GetConfigurations_ is our way to get the configurations we want to build. We'll use a generator/iterator pattern here that we'll see later.
+
+We also need some methods to have callbacks into various stages of the process so that we can do nuget commands and prepare the nuget package.
+
+```python
+
+class BinaryBuildSettingsManager():
+    ''' Platform settings will be accessed through this implementation. '''
+
+    ....
+
+     def PreFirstBuildHook(self):
+        ''' Called after the before the first build '''
+        return 0
+
+    def PostFinalBuildHook(self, ret):
+        ''' Called after the final build with the summed return code '''
+        return 0
+
+    def PostBuildHook(self, ret):
+        ''' Called after each build with the return code '''
+        return 0
+
+    def PreBuildHook(self):
+        ''' Called before each build '''
+        return 0
+```
+
+These hooks are pretty self evident but they'll be called at various point in the process.
+Now with our current file, we can define a settings file that implements the settings class.
+That doesn't really net us much.
+
+## The invocable
+
+The invocable is actually the simplist part of this
+
+```python
+class Edk2BinaryBuild(Edk2Invocable):
+    def GetLoggingLevel(self, loggerType):
+        ''' Get the logging level for a given type
+        base == lowest logging level supported
+        con  == Screen logging
+        txt  == plain text file logging
+        md   == markdown file logging
+        '''
+        if(loggerType == "con") and not self.Verbose:
+            return logging.WARNING
+        return logging.DEBUG
+
+    def AddCommandLineOptions(self, parser):
+        pass
+
+    def RetrieveCommandLineOptions(self, args):
+        '''  Retrieve command line options from the argparser '''
+        pass
+
+    def GetSettingsClass(self):
+        return BinaryBuildSettingsManager
+
+    def GetLoggingFileName(self, loggerType):
+        return "BINARY_BUILDLOG"
```

Comment:
Maybe a comment for the code rather than doc but i would suggest calling the setting manager and getting the build name and using that here too.

Reply:
I was making it simpler for the tutorial's sake